clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java

```diff
@@ -534,15 +594,37 @@ public void onClosed(Consumer<Exception> callback) {
     }
 
     private class ConnectionState implements InvocationBinder {
-        HubConnection connection;
+        private HubConnection connection;
+        private AtomicInteger nextId = new AtomicInteger(0);
+        private HashMap<String, InvocationRequest> pendingInvocations = new HashMap<>();
 
         public ConnectionState(HubConnection connection) {
             this.connection = connection;
         }
 
+        public String getNextId() {
+            int i = nextId.incrementAndGet();
+            return Integer.toString(i);
+        }
+
+        public void cancelOutstandingInvocations(Exception ex) {
+            pendingInvocations.forEach((k, v) -> {
+                if (ex == null) {
+                    v.dispose();
```

Comment:
When you dispose an invocation what happens? Does the caller see an exception?

Reply:
https://github.com/aspnet/SignalR/pull/2961/files#diff-5ec328a725285659deb33f61428b2a70R420
The caller gets a `CancellationException`